File: ts/executor.ts
import getStdin from 'get-stdin'
import minimist from 'minimist'
import { parse } from 'meriyah'
import { generate } from 'astring'

import fs from 'fs'
import {
  Article,
  CodeChunk,
  codeError,
  CodeError,
  CodeExpression,
  EnumSchema,
  Function,
  function_,
  parameter,
  Parameter,
  variable,
  Variable
} from './types'
import { isA } from './util'
import {
  Node as EtreeNode, // In case of conflict with Stencila Node
  BinaryExpression,
  CallExpression,
  FunctionDeclaration,
  FunctionExpression,
  MemberExpression,
  UpdateExpression,
  VariableDeclaration,
  ExpressionStatement,
  AssignmentExpression,
  ImportDeclaration,
  WhileStatement,
  DoWhileStatement,
  IfStatement,
  Program,
  TryStatement
} from 'meriyah/dist/estree'

// eslint-disable-next-line @typescript-eslint/no-floating-promises
main()

interface StringDict {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [key: string]: any
}

interface CliArgs {
  inputFile: string

  outputFile: string

  parameterValues: StringDict
}

interface CodeChunkExecution {
  codeChunk: CodeChunk
  parseResult: CodeChunkParseResult
}

type ExecutableCode = CodeChunkExecution | CodeExpression

/**
 * Process argv to get the source and destination for the document being executed
 */
function getCliArgs(): CliArgs {
  const { _, ...params } = minimist(process.argv.slice(2), {})

  let inFile = '-'
  let outFile = '-'

  if (_.length === 0) {
    // read from stdin and output to stdout
  }

  if (_.length >= 1) {
    // read from provided file arg which might be -, output to stdout
    inFile = _[0]

    if (_.length >= 2) {
      // read from file, out to file. but they might be -
      outFile = _[1]
    }
  }

  return {
    inputFile: inFile,
    outputFile: outFile,
    parameterValues: params
  }
}

/**
 * Execute each item in `code`, making each `parameter` available in the execution context scope.
 */
function execute(code: ExecutableCode[], parameterValues: StringDict): void {
  Object.entries(parameterValues).forEach(([key, value]) => {
    // Add each parameter into the global scope
    // @ts-ignore
    global[key] = value
  })

  code.forEach(c => {
    if (isA('CodeExpression', c)) executeCodeExpression(c)
    else executeCodeChunk(c)
  })
}

/**
 * Traverse a `Node` hierarchy and push any `Parmamater` that is found onto the `paramaters` array,
 * and any executable code block onto the `code` array.
 */
function parseItem(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  item: any,
  parameters: Parameter[],
  code: (CodeChunkExecution | CodeExpression)[],
  functionDepth: number = 0
): void {
  if (isA('Entity', item) || item instanceof Object) {
    if (isA('Function', item)) {
      ++functionDepth
    } else if (isA('Parameter', item) && functionDepth === 0) {
      parameters.push(item)
    } else if (
      (isA('CodeChunk', item) || isA('CodeExpression', item)) &&
      item.language === 'javascript'
    ) {
      if (isA('CodeChunk', item)) {
        const parseResult = parseCodeChunk(item)
        parseResult.finalize()

        item.imports = parseResult.imports
        item.declares = parseResult.declares
        item.assigns = parseResult.assigns
        item.alters = parseResult.alters
        item.uses = parseResult.uses

        if (parseResult.errors.length > 0) {
          if (item.errors === undefined) {
            item.errors = []
          }
          item.errors = item.errors.concat(parseResult.errors)
        }
        code.push({
          codeChunk: item,
          parseResult: parseResult
        })
      } else code.push(item)
    }

    Object.entries(item).forEach(([, i]) => {
      parseItem(i, parameters, code, functionDepth)
    })

    if (isA('Function', item)) {
      --functionDepth
    }
  } else if (Array.isArray(item)) {
    item.forEach(i => parseItem(i, parameters, code, functionDepth))
  }
}

class CodeChunkParseResult {
  public chunkAst: Program | null

  public imports: string[] = []

  public declares: (Variable | Function)[] = []

  public assigns: string[] = []

  public alters: string[] = []

  public uses: string[] = []

  public errors: CodeError[] = []

  private seenIdentifiers: string[] = []

  private possibleVariables: string[] = []

  public constructor(chunkAst: Program | null) {
    this.chunkAst = chunkAst
  }

  private isNameSet(name: string): boolean {
    if (!this.seenIdentifiers.includes(name)) {
      this.seenIdentifiers.push(name)
      return false
    }
    return true
  }

  public addImports(n: string): void {
    if (!this.imports.includes(n)) this.imports.push(n)
  }

  public addPossibleVariable(n: string): void {
    /* If a declaration has a more complex initializer (like a ternary) then we don't know straight away if it's a
     * variable or a function. So the identifier goes in here and if it wasn't computed to be a function declaration
     * after parsing the rest of the code chunk, then set is as a variable.
     */
    if (!this.possibleVariables.includes(n)) this.possibleVariables.push(n)
  }

  public addDeclares(d: Variable | Function): void {
    if (!this.isNameSet(d.name)) this.declares.push(d)
  }

  public addAssigns(n: string): void {
    if (!this.isNameSet(n)) this.assigns.push(n)
  }

  public addAlters(n: string): void {
    if (!this.isNameSet(n)) this.alters.push(n)
  }

  public addUses(n: string): void {
    if (!this.isNameSet(n)) this.uses.push(n)
  }

  public finalize(): void {
    this.possibleVariables.forEach(n => {
      if (!this.seenIdentifiers.includes(n)) this.declares.push(variable(n))
    })
  }
}

function recurseMemberExpression(expr: MemberExpression): string | null {
  if (expr.object.type === 'Identifier') return expr.object.name
  if (expr.object.type === 'MemberExpression')
    return recurseMemberExpression(expr.object)

  return null
}

function parseFunctionDeclaration(
  result: CodeChunkParseResult,
  fn: FunctionDeclaration | FunctionExpression,
  name?: string | null
): void {
  if (name === undefined) {
    name = fn.id !== null ? fn.id.name : null
  }

  if (name === null) return

  const parameters: Parameter[] = []

  if (fn.params !== undefined) {
    fn.params.forEach(p => {
      if (p.type === 'Identifier') parameters.push(parameter(p.name))
      else if (p.type === 'RestElement' && p.argument.type === 'Identifier')
        parameters.push(parameter(p.argument.name, { repeats: true }))
    })
  }

  result.addDeclares(function_(name, { parameters }))
}

function parseVariableDeclaration(
  result: CodeChunkParseResult,
  statement: VariableDeclaration
): void {
  statement.declarations.forEach(declarator => {
    if (declarator.id.type === 'Identifier') {
      if (declarator.init === null || declarator.init.type === 'Literal')
        result.addDeclares(variable(declarator.id.name))
      else {
        result.addPossibleVariable(declarator.id.name)
        parseStatement(result, declarator.init, declarator.id.name)
      }
    }
  })
}

function parseAssignmentExpression(
  result: CodeChunkParseResult,
  statement: AssignmentExpression
): void {
  let assignmentName: string | undefined

  if (statement.left.type === 'Identifier') {
    assignmentName = statement.left.name
    result.addAssigns(statement.left.name)
  } else if (statement.left.type === 'MemberExpression') {
    const name = recurseMemberExpression(statement.left)
    if (name !== null) result.addAlters(name)
  }

  /* an offshoot of only setting assignmentName for Identifiers only (not MemberExpressions) is that functions that
  are declared and assigned to a property of an object won't be parsed, since their name will be like `a.b` so won't be
  a valid function identifier
  */
  parseStatement(result, statement.right, assignmentName)
}

function parseBinaryExpression(
  result: CodeChunkParseResult,
  statement: BinaryExpression
): void {
  if (statement.left.type === 'Identifier') result.addUses(statement.left.name)
  else parseStatement(result, statement.left)

  if (statement.right.type === 'Identifier')
    result.addUses(statement.right.name)
  else parseStatement(result, statement.right)
}

function parseCallExpression(
  result: CodeChunkParseResult,
  statement: CallExpression
): void {
  statement.arguments.forEach(arg => {
    if (arg.type === 'Identifier') result.addUses(arg.name)
    else parseStatement(result, arg)
  })
}

function parseUpdateExpression(
  result: CodeChunkParseResult,
  statement: UpdateExpression
): void {
  if (statement.argument.type === 'Identifier')
    result.addAlters(statement.argument.name)
  else parseStatement(result, statement.argument)
}

function parseImportExpression(
  result: CodeChunkParseResult,
  statement: ImportDeclaration
): void {
  if (
    statement.source.type === 'Literal' &&
    typeof statement.source.value === 'string'
  ) {
    result.addImports(statement.source.value)
  }
}

function parseWhileStatement(
  result: CodeChunkParseResult,
  statement: WhileStatement | DoWhileStatement
): void {
  parseStatement(result, statement.test)

  parseStatement(result, statement.body)
}

function parseIfStatement(
  result: CodeChunkParseResult,
  statement: IfStatement
): void {
  parseStatement(result, statement.test)
  parseStatement(result, statement.consequent)
  if (statement.alternate !== null) {
    parseStatement(result, statement.alternate)
  }
}

function parseTryStatement(
  result: CodeChunkParseResult,
  statement: TryStatement
): void {
  statement.block.body.forEach(subStatement => {
    parseStatement(result, subStatement)
  })

  if (statement.handler !== null) {
    statement.handler.body.body.forEach(subStatement => {
      parseStatement(result, subStatement)
    })
  }

  if (statement.finalizer !== null) {
    statement.finalizer.body.forEach(subStatement => {
      parseStatement(result, subStatement)
    })
  }
}

function parseExpression(
  result: CodeChunkParseResult,
  statement: ExpressionStatement
): void {
  parseStatement(result, statement.expression)
}

function parseStatement(
  result: CodeChunkParseResult,
  statement: EtreeNode,
  lastParsedVarName?: string
): void {
  switch (statement.type) {
    case 'VariableDeclaration':
      parseVariableDeclaration(result, statement as VariableDeclaration)
      break
    case 'ExpressionStatement':
      parseExpression(result, statement as ExpressionStatement)
      break
    case 'AssignmentExpression':
      parseAssignmentExpression(result, statement as AssignmentExpression)
      break
    case 'BinaryExpression':
      parseBinaryExpression(result, statement as BinaryExpression)
      break
    case 'FunctionDeclaration':
      parseFunctionDeclaration(result, statement as FunctionDeclaration)
      break
    case 'FunctionExpression':
      parseFunctionDeclaration(
        result,
        statement as FunctionExpression,
        lastParsedVarName
      )
      break
    case 'CallExpression':
      parseCallExpression(result, statement as CallExpression)
      break
    case 'UpdateExpression':
      parseUpdateExpression(result, statement as UpdateExpression)
      break
    case 'ImportDeclaration':
      parseImportExpression(result, statement as ImportDeclaration)
      break
    case 'WhileStatement':
    case 'DoWhileStatement':
      // both actually have the same interface so it is OK to cast just to WhileStatement
      parseWhileStatement(result, statement as WhileStatement)
      break
    case 'IfStatement':
    case 'ConditionalExpression':
      // both actually have the same interface so it is OK to cast just to IfStatement
      parseIfStatement(result, statement as IfStatement)
      break
    case 'BlockStatement':
      statement.body.forEach(subStatement => {
        parseStatement(result, subStatement)
      })
      break
    case 'TryStatement':
      parseTryStatement(result, statement)
      break
    case 'EmptyStatement':
    case 'Identifier':
    case 'UnaryExpression':
    case 'Literal':
    case 'ThrowStatement':
      break
    default:
      console.log(statement)
      throw new Error(`Unhandled statement ${statement.type}`)
  }
}

function exceptionToCodeError(error: Error | string): CodeError {
  if (typeof error === 'string') {
    return codeError('Exception', { message: error })
  } else {
    return codeError(error.name, { message: error.message, trace: error.stack })
  }
}

function setCodeError(
  code: CodeChunk | CodeExpression,
  error: Error | string
): void {
  if (code.errors === undefined) {
    code.errors = []
  }
  code.errors.push(exceptionToCodeError(error))
}

function parseCodeChunk(codeChunk: CodeChunk): CodeChunkParseResult {
  let chunkAst: Program | null = null

  try {
    chunkAst = parse(codeChunk.text, { module: true })
  } catch (e) {
    const badParseResult: CodeChunkParseResult = new CodeChunkParseResult(null)
    badParseResult.errors.push(exceptionToCodeError(e))
    return badParseResult
  }

  const parseResult: CodeChunkParseResult = new CodeChunkParseResult(chunkAst)

  chunkAst.body.forEach(statement => {
    parseStatement(parseResult, statement)
  })

  return parseResult
}

/**
 * Execute a `CodeExpression` (which should be a single expression) and set the result onto its `output` property.
 *
 * Uses `eval`, so should only be called with trusted code.
 */
function executeCodeExpression(code: CodeExpression): void {
  try {
    // eslint-disable-next-line no-eval
    code.output = eval(code.text)
  } catch (e) {
    setCodeError(code, e)
  }
}

/**
 * Execute a `CodeChunk` (which might be multiple lines) and set any return values or console output as items in an
 * array on it `outputs` property.
 *
 * Uses `eval`, so should only be called with trusted code.
 */
function executeCodeChunk(code: CodeChunkExecution): void {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const outputs: any[] = []

  const ast = code.parseResult.chunkAst

  if (ast === null) return

  const chunk = code.codeChunk

  ast.body.forEach(statement => {
    /*
    Convert `const` or `let` global declarations to `var`, otherwise they are lost in subsequent eval() calls due to
    scope changes
     */
    if (statement.type === 'VariableDeclaration' && statement.kind !== 'var')
      statement.kind = 'var'

    // @ts-ignore
    const generatedCode = generate(statement)

    let loggedData = ''

    const oldCl = console.log

    console.log = (s: string) => {
      loggedData += s
    }

    let res

    try {
      // @ts-ignore
      // eslint-disable-next-line no-eval
      res = (1, eval)(generatedCode)
    } catch (e) {
      setCodeError(chunk, e)
    }

    console.log = oldCl

    outputs.push(res)

    if (loggedData.length > 0) {
      outputs.push(loggedData)
    }
  })

  chunk.outputs = outputs.filter(o => o !== undefined)
}

/**
 * Decode a parameter (e.g. read from the command line, a string) into a value based on the `schema` of the `parameter`.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function decodeParameter(parameter: Parameter, value?: string): any {
  if (isA('ConstantSchema', parameter.schema)) return parameter.schema.value

  if (value === undefined) {
    if (parameter.default === undefined)
      throw new Error(
        `No value or default found for parameter ${parameter.name}`
      )

    return parameter.default
  }

  if (parameter.schema === undefined) return value

  switch (parameter.schema.type) {
    case 'ArraySchema':
    case 'TupleSchema':
      return JSON.parse(value)
    case 'BooleanSchema':
      const b = value.toLowerCase()
      return b === 'true' || b === 't' || b === 'yes' || b === '1'
    case 'EnumSchema':
      const es = parameter.schema as EnumSchema
      if (es.values !== undefined && !es.values.includes(value))
        throw new Error(
          `${value} not found in enum values for ${parameter.name}`
        )
      return value
    case 'IntegerSchema':
      const intValue = parseInt(value)
      if (isNaN(intValue)) throw new Error(`$[raw} is not a valid integer`)
      return intValue
    case 'NumberSchema':
      const floatValue = parseFloat(value)
      if (isNaN(floatValue)) throw new Error(`$[raw} is not a valid float`)
      return floatValue
  }

  return value
}

/**
 * Decode all the parameters read from the command line, based on the `Parameter` nodes found in the document.
 */
function decodeParameters(
  parameters: Parameter[],
  values: { [key: string]: string }
): StringDict {
  const decodedValues: StringDict = {}

  parameters.forEach(p => {
    decodedValues[p.name] = decodeParameter(p, values[p.name])
  })

  return decodedValues
}

/**
 * Read the input document from a file or stdin.
 */
async function readInput(path: string): Promise<string> {
  if (path === '-') {
    return getStdin()
  }

  return fs.readFileSync(path, 'utf8')
}

/**
 * Write the executed document (`Article`) to a file or stdout.
 */
function outputArticle(path: string, output: Article): void {
  const j = JSON.stringify(output, null, 2)

  if (path === '-') {
    console.log(j)
  } else {
    fs.writeFileSync(path, j)
  }
}

/**
 * Execute a document based on arguments from the command line.
 */
async function main(): Promise<void> {
  const cliArgs = getCliArgs()
  const article = JSON.parse(await readInput(cliArgs.inputFile))

  if (!isA('Article', article)) throw TypeError('Not an Article')

  const parameters: Parameter[] = []
  const code: ExecutableCode[] = []

  parseItem(article, parameters, code)

  execute(code, decodeParameters(parameters, cliArgs.parameterValues))

  outputArticle(cliArgs.outputFile, article)
}

File: py/executor.py
import argparse
import ast
import base64
import json
import logging
import sys
import typing
from contextlib import redirect_stdout
from io import TextIOWrapper, BytesIO

import astor
from stencila.schema.types import Parameter, CodeChunk, Article, Entity, CodeExpression, ConstantSchema, EnumSchema, \
    BooleanSchema, NumberSchema, IntegerSchema, StringSchema, ArraySchema, TupleSchema, ImageObject, Datatable, \
    DatatableColumn
from stencila.schema.util import from_json, to_json

try:
    import matplotlib.figure
    import matplotlib.artist

    MPLFigure = matplotlib.figure.Figure
    MPLArtist = matplotlib.artist.Artist
    mpl_available = True
except ImportError:
    class MPLFigure:
        pass


    class MLPArtist:
        pass


    mpl_available = False

try:
    from pandas import DataFrame
    import numpy

    pandas_available = True
except ImportError:
    class DataFrame:
        pass


    pandas_available = False

logger = logging.getLogger(__name__)
logger.addHandler(logging.NullHandler())

ExecutableCode = typing.Union[CodeChunk, CodeExpression]


class StdoutBuffer(TextIOWrapper):
    def write(self, string: typing.Union[bytes, str]) -> int:
        if isinstance(string, str):
            return super(StdoutBuffer, self).write(string)
        else:
            return super(StdoutBuffer, self).buffer.write(string)


class DocumentParser:
    """Parse an executable document (`Article`) and cache references to its parameters and code nodes."""

    parameters: typing.List[Parameter] = []
    code: typing.List[ExecutableCode] = []

    def parse(self, source: Article) -> None:
        # todo: this traverses the article twice. Make it less hard coded, maybe pass through a lookup table that maps
        # a found type to its destination
        self.handle_item(source, Parameter, self.parameters, None)
        self.handle_item(source, (CodeChunk, CodeExpression), self.code, {'language': 'python'})

    def handle_item(self, item: typing.Any,
                    search_type: typing.Union[typing.Type[Entity], typing.Iterable[typing.Type[Entity]]],
                    destination: typing.List[Entity],
                    attr_match: typing.Optional[typing.Dict[str, typing.Any]]) -> None:
        if isinstance(item, dict):
            self.traverse_dict(item, search_type, destination, attr_match)
        elif isinstance(item, list):
            self.traverse_list(item, search_type, destination, attr_match)
        elif isinstance(item, Entity):
            if isinstance(item, search_type):
                can_add = True
                if attr_match:
                    for k, v in attr_match.items():
                        if getattr(item, k, None) != v:
                            can_add = False
                            break
                if can_add:
                    logger.debug('Adding {}'.format(type(item)))
                    destination.append(item)
            self.traverse_dict(item.__dict__, search_type, destination, attr_match)

    def traverse_dict(self, d: dict,
                      search_type: typing.Union[typing.Type[Entity], typing.Iterable[typing.Type[Entity]]],
                      destination: typing.List[Entity],
                      attr_match: typing.Optional[typing.Dict[str, typing.Any]]) -> None:
        for child in d.values():
            self.handle_item(child, search_type, destination, attr_match)

    def traverse_list(self, l: typing.List,
                      search_type: typing.Union[typing.Type[Entity], typing.Iterable[typing.Type[Entity]]],
                      destination: typing.List[Entity],
                      attr_match: typing.Optional[typing.Dict[str, typing.Any]]) -> None:
        for child in l:
            self.handle_item(child, search_type, destination, attr_match)


class Executor:
    """Execute a list of code blocks, maintaining its own `globals` scope for this execution run."""

    globals: typing.Optional[typing.Dict[str, typing.Any]]

    def execute_code_chunk(self, chunk: CodeChunk, _locals: typing.Dict[str, typing.Any]) -> None:
        cc_outputs = []

        tree = ast.parse(chunk.text, 'exec')

        for statement in tree.body:
            capture_result = False

            if isinstance(statement, ast.Expr):
                capture_result = True
                run_function = eval
                code_to_run = astor.to_source(statement)
            else:
                run_function = exec
                m = ast.Module()
                m.body = [statement]
                code_to_run = compile(m, '<ast>', 'exec')

            s = StdoutBuffer(BytesIO(), sys.stdout.encoding)

            with redirect_stdout(s):
                result = run_function(code_to_run, self.globals, _locals)

            if capture_result and result is not None:
                cc_outputs.append(self.decode_output(result))

            s.seek(0)
            std_out_output = s.buffer.read()

            if std_out_output:
                cc_outputs.append(std_out_output.decode('utf8'))

        chunk.outputs = cc_outputs

    def execute_code_expression(self, expression: CodeExpression, _locals: typing.Dict[str, typing.Any]) -> None:
        expression.output = self.decode_output(eval(expression.text, self.globals, _locals))

    def execute(self, code: typing.List[ExecutableCode], parameter_values: typing.Dict[str, typing.Any]) -> None:
        self.globals = {}

        _locals = parameter_values.copy()

        for c in code:
            if isinstance(c, CodeChunk):
                self.execute_code_chunk(c, _locals)
            elif isinstance(c, CodeExpression):
                self.execute_code_expression(c, _locals)
            else:
                raise TypeError('Unknown Code node type found: {}'.format(c))

    @staticmethod
    def value_is_mpl(value: typing.Any) -> bool:
        if not mpl_available:
            return False

        return isinstance(value, (MPLFigure, MPLArtist)) or (
                isinstance(value, list) and len(value) == 1 and isinstance(value[0], MPLArtist))

    @staticmethod
    def decode_mpl() -> ImageObject:
        image = BytesIO()
        matplotlib.pyplot.savefig(image, format='png')
        src = 'data:image/png;base64,' + base64.encodebytes(image.getvalue()).decode()
        return ImageObject(src)

    @staticmethod
    def decode_dataframe(df: DataFrame) -> Datatable:
        columns = []

        for column_name in df.columns:
            column = df[column_name]
            values = column.tolist()
            if column.dtype in (numpy.bool_, numpy.bool8):
                schema = BooleanSchema()
                values = [bool(row) for row in values]
            elif column.dtype in (numpy.int8, numpy.int16, numpy.int32, numpy.int64):
                schema = IntegerSchema()
                values = [int(row) for row in values]
            elif column.dtype in (numpy.float16, numpy.float32, numpy.float64):
                schema = NumberSchema()
                values = [float(row) for row in values]
            elif column.dtype in (numpy.str_, numpy.unicode_,):
                schema = StringSchema()
            else:
                schema = None

            columns.append(
                DatatableColumn(column_name, values, schema=ArraySchema(items=schema))
            )

        return Datatable(columns)

    def decode_output(self, output: typing.Any) -> typing.Any:
        if self.value_is_mpl(output):
            return self.decode_mpl()

        if isinstance(output, DataFrame):
            return self.decode_dataframe(output)

        return output


class ParameterParser:
    """
    Parse parameters that the document requires, from the command line.

    The `ArgumentParser` will fail if any required parameters are not passed in.
    """
    parameters: typing.Dict[str, Parameter]
    parameter_values: typing.Dict[str, typing.Any]

    def __init__(self, parameters: typing.List[Parameter]) -> None:
        self.parameters = {parameter.name: parameter for parameter in parameters}
        self.parameter_values = {}

    def parse_cli_args(self, cli_args: typing.List[str]) -> None:
        if not self.parameters:
            logger.debug('No parameters passed to parse_cli_args')
            return

        param_parser = argparse.ArgumentParser(description='Parse Parameters')

        for param in self.parameters.values():
            if not isinstance(param.schema, ConstantSchema):
                param_parser.add_argument('--' + param.name, dest=param.name,
                                          required=self.get_parameter_default(param) is None)

        args, _ = param_parser.parse_known_args(cli_args)

        logger.debug('Parsed command line args: {}'.format(args))

        for param_name in self.parameters:
            cli_value = getattr(args, param_name, None)
            if not cli_value and self.parameters[param_name].default:
                self.parameter_values[param_name] = self.parameters[param_name].default
            else:
                self.parameter_values[param_name] = self.deserialize_parameter(self.parameters[param_name], cli_value)

    @staticmethod
    def get_parameter_default(parameter: Parameter) -> typing.Any:
        if isinstance(parameter.schema, ConstantSchema):
            return parameter.schema.value or parameter.default

        return parameter.default

    @staticmethod
    def deserialize_parameter(parameter: Parameter, value: typing.Any) -> typing.Any:
        # Lots of TODOs here, might not care as passing this off to encoda soon

        if isinstance(parameter.schema, ConstantSchema):
            return parameter.schema.value

        if isinstance(parameter.schema, EnumSchema):
            if value not in parameter.schema.values:
                raise TypeError('{} not found in enum values for {}'.format(value, parameter.name))
            return value

        if isinstance(parameter.schema, BooleanSchema):
            return value.lower() in ('true', 'yes', '1' 't')

        if isinstance(parameter.schema, IntegerSchema):
            return int(value)

        if isinstance(parameter.schema, NumberSchema):
            return float(value)  # TODO should be a decimal?

        if isinstance(parameter.schema, StringSchema):
            return value

        if isinstance(parameter.schema, ArraySchema):
            return json.loads(value)

        if isinstance(parameter.schema, TupleSchema):
            return json.loads(value)


def execute_document(cli_args: typing.List[str]):
    cli_parser = argparse.ArgumentParser()
    cli_parser.add_argument('input_file', help='File to read from or "-" to read from stdin', nargs='?', default='-')
    cli_parser.add_argument('output_file', help='File to write to or "-" to write to stdout', nargs='?', default='-')

    args, _ = cli_parser.parse_known_args(cli_args)

    if args.input_file == '-':
        j = sys.stdin.read()
    else:
        with open(args.input_file) as i:
            j = i.read()

    article = from_json(j)

    if not isinstance(article, Article):
        raise TypeError('Decoded JSON was not an Article')

    doc_parser = DocumentParser()
    doc_parser.parse(article)

    e = Executor()

    pp = ParameterParser(doc_parser.parameters)
    pp.parse_cli_args(cli_args)

    e.execute(doc_parser.code, pp.parameter_values)

    if args.output_file == '-':
        sys.stdout.write(to_json(article))
    else:
        with open(args.output_file, 'w') as o:
            o.write(to_json(article))


if __name__ == '__main__':
    logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)
    execute_document(sys.argv[1:])
